### DISCUSSION TITLE
mxnet 2.0 / gluon questions

### DISCUSSION THREAD

**Discussion by feevos** [apache/incubator-mxnet#19109] [Q&A] [answered]
Dear all, 

with the new version of mxnet 2.0 coming up, I have several questions as the documentation in the official site is not always clear (or I fail to understand) if it refers to mxnet 2.0 or mxnet 1.XX (even if in the drop down menu I choose master). In particular, I have the following two questions: 

1. How will hybridization work in the new mxnet? Will the hybrid_forward function be non transparent to the user (so the user just needs to define forward, with an np. operator, and then call hybridize)? Or users should still use the ```hybrid_forward(self, F, input)```? Could you please provide an example of a custom layer (say conv2D + batchnorm) with the new mxnet+gluon 2.0 API that is hybridizable?
2. I think there is a depreciation of the ```self.name_scope()```, but it is not clear to me if this takes place only in mxnet2.0 or in earlier versions too (1.7). 

In summary, if you could please provide a simple example of a custom (hybridizable) operator for conv2d + batchnorm that would explain a lot.

Many thanks,
Foivos

**Comment by szha** [comment, marked as answer]
Hi @feevos,

Thanks for asking these questions. It helps us identify gaps in our communication.

> 1. How will hybridization work in the new mxnet?

The preferred way of hybridization will be with the `def forward(self, x, ...)` signature without implicit parameter passing. The hybridization will work by tracing the NDArrays directly without using the placeholder `F` for python-side tracing. This enables more expressiveness in the models that can be hybridized.

> 2. I think there is a depreciation of the self.name_scope()

Deprecation only happens in 2.0 though it's not necessary anymore in later versions of 1.x. User code with `name_scope()` will continue to work on v1.x versions.

Once we get the website into a good shape, we will make sure to mention these in an upgrade guide. Let me know if you have more question.

Best,
Sheng

**Reply by feevos** [reply]
Thank you very much!
